### PR TITLE
[ML] Exclude features which are unlikely to be selected during training

### PR DESCRIPTION
**Pull request by tveasey**
This represents a better runtime QoR tradeoff since such features typically aren't selected in the tree, but cost training runtime. This also defends us against an edge case where the user provides a very large number of features.

We set the threshold on the probability to select fairly low by default, but this provides a useful knob which we can use to speed up training on large data sets in the future. I also took the opportunity to switch to the named parameter idiom for `CDataFrameCategoryEncoder` because the constructor was getting a bit unwieldy.

Finally, we should have been basing the maximum tree size on the number of rows of training data and not overall number of rows.

### CODE REVIEW

File: include/maths/CBoostedTreeImpl.h [hunk 1]
```diff
@@ -735,7 +736,7 @@ class MATHS_EXPORT CBoostedTreeImpl final {
     //!
     //! \note This number will only be used if the regularised loss says its
     //! a good idea.
-    std::size_t maximumTreeSize(const core::CDataFrame& frame) const;
+    std::size_t maximumTreeSize(const core::CPackedBitVector& frame) const;
```
Comment (hendrikmuhs) [inline include/maths/CBoostedTreeImpl.h]: nit: the parameter seems not to be a (data) frame anymore, impl uses `trainingRowMask`

File: include/maths/CDataFrameCategoryEncoder.h [hunk 2]
```diff
@@ -68,6 +70,8 @@ class MATHS_EXPORT CEncodedDataFrameRowRef final {
     const CDataFrameCategoryEncoder* m_Encoder;
 };
 
+class CMakeDataFrameCategoryEncoder;
```
[the remark this answers is not in the document]
Comment (tveasey) [reply include/maths/CDataFrameCategoryEncoder.h]: I know that strictly these are a wrapper around the encoder parameters, but the reason I didn't do this is I think it makes the code where this is used easier to read IMO. At the moment you can write something like:
```
CDataFrameCategoryEncoder encoder =
            CMakeDataFrameCategoryEncoder{threads, frame, dependentVariable}
            .minimumRowsPerFeature(rowsPerFeature)
            ....
```
which reads better to me, for example compare with say `OpenFile` or some other likely candidate for this idiom.

File: include/maths/CBoostedTreeImpl.h (hunk 1 above)
Comment (tveasey) [reply include/maths/CBoostedTreeImpl.h]: [bf51751](https://github.com/elastic/ml-cpp/pull/640/commits/bf51751decfd8c7f8aab367652fb87c07bd29667)

[the remark this answers is not in the document]
Comment (tveasey) [reply include/maths/CBoostedTreeImpl.h]: [bf51751](https://github.com/elastic/ml-cpp/pull/640/commits/bf51751decfd8c7f8aab367652fb87c07bd29667)

**Pull request merged**